### ISSUE TITLE
[Chrome] CSP worker-src hash does not work

### ISSUE THREAD

**Issue by t2ym** [t2ym/thin-hook#268]
CSP `worker-src 'hash'` directive does not work

As of Chrome 70,

### The following directive passes all the worker scripts in the same origin, i.e., the hash is ineffective.
`worker-src 'sha256-iG93ORbzC6MgtFIuAfnoxiH35oIEKWDvCrl3Ykyft8U=' 'self'`

### The following directive denies all the worker scripts, i.e., the hash is ineffective.
`worker-src 'sha256-iG93ORbzC6MgtFIuAfnoxiH35oIEKWDvCrl3Ykyft8U='`

### Error Message with hash-only `worker-src` directive
"Refused to create a worker from '{worker URL}' because it violates the following Content Security Policy directive: "worker-src 'sha256-iG93ORbzC6MgtFIuAfnoxiH35oIEKWDvCrl3Ykyft8U='"."

### `require-sri-for script` directive with WebPlatformFeatures-enabled spits out unintelligible error message as below, since Service Workers have no attributes like `<script>` elements.
"Refused to load the service worker 'URL...' because 'require-sri-for' directive requires integrity attribute be present for all service workers."

## Suspected Root Cause
- Implementation is still in progress for `worker-src` directive.
[Chromium bug 662930: CSP3: Implement worker changes](https://bugs.chromium.org/p/chromium/issues/detail?id=662930&can=1&q=worker-src&colspec=ID%20Pri%20M%20Stars%20ReleaseBlock%20Component%20Status%20Owner%20Summary%20OS%20Modified)
- CSP3 Editor's Draft seems to be incomplete
  - [6.1.17.1. worker-src Pre-request Check](https://w3c.github.io/webappsec-csp/#worker-src-pre-request)
  - [6.1.17.2. worker-src Post-request Check](https://w3c.github.io/webappsec-csp/#worker-src-post-request)
    - It seems to be impossible to define `worker-src 'hash'` directives that pass Pre-request Check but denies Post-request Check.
    - It seems that `worker-src 'hash'` directives should be applied only at Post-request check and skips Pre-request Check so that hashes of the target worker scripts can be checked at post-request check.
    - From the Chromium sources, it seems that only the Pre-request Check on `worker-src` directive is implemented.

**Comment by t2ym** [comment]
## An alternative solution is to be verified. Can this be a breakthrough?
- Using only the current standard and implemented API
- Applicable to Service Worker

## Details to be announced only after the implementation becomes stable and complete

**Comment by thomas-panteum** [comment]
Hi guys!

Any news about this?
We are waiting badly for this feature as worker-src sha256 directive still doesn't work even if the hash is valid.

**Comment by t2ym** [comment]
Hi Thomas @thomas-panteum,

From the CSP3 specs and the Chromium implementation, I have an impression that CSP3 does not expect worker-src sha256 directive while the syntax for HTTP headers is defined clearly. So the feature may be just out of scope.

My current workaround solution in `thin-hook/demo/integrity.js` is as follows.

1. Make the Service Worker script (`hook.min.js`) hybrid, which detects whether it is running in the main document or as a Service Worker
2. Add a valid `integrity` attribute value to the script tag in the main document
3. Treat as a fatal error if the script does not load
4. Load the Service Worker from the same URL path as one in the main document
5. Check the Service Worker integrity by fetching it from the main document with these options: `cache: only-if-cached` and `integrity: integrity`
```js
let request = new Request(serviceWorkerPath, { headers: headers, mode: 'same-origin', cache: 'only-if-cached', integrity: integrity });
```
6. If the fetching fails, the web app will terminate (redirect to `about:blank`) due to the fatal error.
7. Detect `updatefound` events of the Service Worker and the web app will terminate if the updated Service Worker is for the same or unknown app version (`hook.min.js?version={version_XXX}`)

**This workaround assumes that the Service Worker script cached in the browser and fetched from the cache is the byte-by-byte same as the one which has just been loaded.**

It is robust as long as `integrity.js` carefully confirms integrity of the scripts and the main document step by step.

### Note:
- This `only-if-cached` option is not compatible with Chrome for Android.

### P.S.
While your comment starts with "Hi guys!" (plural), I am the only and lonely contributor for this project and you are the very first GitHub user who has made a comment on this `thin-hook` project.  I definitely need productive feedbacks from users and I am really happy to respond to them like yours. Thank you.

**Comment by t2ym** [comment]
With 0.4.0-alpha.2, `integrity.js` is updated to avoid replacing the entry page after a hacked Service Worker script is loaded.

**Comment by t2ym** [comment]
**With 0.4.0-alpha.1**, the following MITM attack in `demo-backend/demoServer.js` can forcefully replace the entry page with a hacked one, which is actually incapable of doing any malicious activities as it cannot break the double encryption #310.

**With 0.4.0-alpha.2**, the entry page hangs up with an infinite loop or transitions to `about:blank` if this MITM attack is detected.

```js
    .get('/components/thin-hook/hook.min.js', (req, res, next) => {
      // mitm attacker for hook.min.js
      let keys = Object.keys(req.headers);
      if (true) {
        console.log(pid + 'req.url = ' + req.url + '\n' + JSON.stringify(req.headers, null, 2) + '\n');
      }
      if (keys.indexOf('user-agent') < keys.indexOf('service-worker')) {
        console.log(pid + 'hook.min.js: real Service Worker request; Sending hacked version');
        const hackedHookMinJs = `
          const forceNavigation = true; // true to navigate to hacked-index.html on activation
          const delayForcedNavigation = 100; // ms; 100ms => Redirect to about:blank, 10ms => Hang up with an infinite loop by integrity.js
          oninstall = function (event) {
            console.log('oninstall', event);
          };
          onactivate = function (event) {
            console.log('onactivate', event);
            event.waitUntil(clients.claim().then(() =>
              clients.matchAll({ includeUncontrolled: true, type: 'window' }).then(function (clientList) {
                console.log('clientList.length', clientList.length);
                for (var i = 0; i < clientList.length; i++) {
                  console.log('clientList[' + i + ']', clientList[i]);
                  if (forceNavigation) {
                    if (delayForcedNavigation > 0) {
                      let client = clientList[i];
                      setTimeout(() => client.navigate('hacked-index.html'), delayForcedNavigation);
                    }
                    else {
                      clientList[i].navigate('hacked-index.html');
                    }
                  }
                }
              })
            ));
          };
          onfetch = function (event) {
            console.log('onfetch', event.request.url);
            event.respondWith(((new URL(event.request.url).pathname) === '/components/thin-hook/hacked-index.html')
              ? new Response(
                '<html lang="en"><head><title>Hacked thin-hook Demo!!!</title></head><body><h1>Hacked thin-hook Demo!!! But the page can do practically nothing.</h1></body></html>',
                { headers: new Headers({ 'Content-Type': 'text/html' }) })
              : fetch(event.request));
          };
          console.log('hacked-hook.min.js executed');`;
        res.setHeader('Service-Worker-Allowed', '/');
        res.setHeader('Content-Type', 'text/javascript');
        res.send(hackedHookMinJs);
      }
      else {
        console.log(pid + 'hook.min.js: non-Service Worker request')
        res.sendFile(path.join(rootPhysicalPath, 'hook.min.js'), {
          headers: {
            //'cache-control': 'max-age=60',
            //'content-security-policy': 'require-sri-for script; script-src \'sha256-aaaa\'',
            //'content-security-policy': 'require-sri-for script; worker-src \'sha256-jG8jnbsyHT6GzdaWaf93W02ykkaRjZA5ttOaR+tpP0Q=\' \'self\'',
            //'content-security-policy': 'require-sri-for script; script-src \'sha256-jG8jnbsyHT6GzdaWaf93W02ykkaRjZA5ttOaR+tpP0Q=\' \'unsafe-eval\'',
          }
        });
      }
    })
```
